### PR TITLE
Enhancements: configurable database file path, default port, and optimised Docker image

### PR DESCRIPTION
**Pull request by MrSaints**
**Configurable database file path**

The sqlite3 database file path is currently hardcoded. It'd be great if it can be configured. For example, if we were to mount a Docker volume for persistence, it'd be far easier if we could mount a data directory rather than a specific file which most implementations do not nicely support, e.g. Rancher. 

**Default port**

I noticed that there's no default set, and the binding can be set to `:`. I have set it to `8080` as per the default environment variable for it.

**Optimised Docker image**

The main goal here is to optimise for caching, build speed and final image size. It takes advantage of a multi-stage Docker build.

- Build backend, and frontend in separate images
- Copy artifacts from build images to final, execution image
- Add `.dockerignore` to speed up build times (and improve caching)
- Copy, and install npm packages before building (again, to improve caching)
- Build Go binary statically

The final size is under 20MB compared to the hundreds.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ func main() {
 	http.HandleFunc("/get", GetCommentsHandler)
 
 	port := os.Getenv("COMMENTO_PORT")
+	if port == "" {
+		port = "8080"
+	}
```
Comment (adtac) [inline main.go]: This is already done: see [config.go:12](https://github.com/adtac/commento/blob/master/config.go#L12).

File: main.go [hunk 2]
```diff
@@ -11,7 +11,12 @@ import (
 func main() {
 	var err error
 
-	err = LoadDatabase("sqlite:file=sqlite3.db")
+	fp := os.Getenv("COMMENTO_DATABASE_FILE")
+	if fp == "" {
+		fp = "sqlite3.db"
+	}
```
Comment (adtac) [inline main.go]: This should be performed in config.go.

File: main.go [hunk 3]
```diff
@@ -11,7 +11,12 @@ import (
...
+
+	err = LoadDatabase("sqlite:file=" + fp)
```
Comment (adtac) [inline main.go]: Directly do `os.Getenv("...")` after you've made the config.go change

File: .dockerignore [hunk 4]
```diff
@@ -0,0 +1,7 @@
+*.db
+*.md
+.dockerfile
+.editorconfig
+.gitignore
+Dockerfile
+example/
```
Comment (adtac) [inline .dockerignore]: newline at EOF please

File: Dockerfile [hunk 5]
```diff
@@ -1,22 +1,33 @@
...
+FROM golang:1.8.1 as backend-build
 
-# copy source
 COPY . /go/src/commento
 WORKDIR /go/src/commento
 
-# build backend
 RUN go get -v .
-RUN go install .
+RUN go build -ldflags '-linkmode external -extldflags -static -w'
```
Comment (adtac) [inline Dockerfile]: Not a review comment, just curious: what does this do? I haven't been writing Go long enough ^^

Searching tells me `-linkmode external` can be used to load go variables at compile time. Is that the reasoning?

`-static` is static linking, I can tell. `-w` is some kind of executable size reduction technique from what I read (add `-s` as well?).

**Comment by adtac** [comment]
(PS: @MrSaints you'll need to use git-rebase to amend the commits; don't use merge commits or add a new commit on top of the three.)

File: Dockerfile (hunk 5 above)
Comment (MrSaints) [reply Dockerfile]: The reason I went with this versus a standard `go build` is because I wanted to build a static binary from the `golang` Docker image, but still be able to use it in `alpine`. There are a couple of problems if we rely on the former.

One of which is your project's use of `mattn/go-sqlite3`. It is a `cgo` package, not a Go package. Your project however, is in Go. So we have situation where we need a combination of Go, and non-Go code to be in the final executable without dynamic libraries. So, `-linkmode external` simply tells `cmd/linker` to invoke the host linker (often `gcc`) to combine your Go code (compiled by Go to an object file `go.o`, but not processed any further) with your non-Go code.

It is also possibly using `cgo` from the `net` library (see https://golang.org/pkg/net/#hdr-Name_Resolution). 

Everything after `-extldflags` are flags for `gcc` (`-w` to suppress warnings IIRC).

TL;DR: I just want a static binary

_(see https://github.com/mattn/go-sqlite3/issues/384#issuecomment-290291449 for more information)_

File: main.go (hunk 1 above)
Comment (MrSaints) [reply main.go]: Ah yes, that's right. For some reason it didn't work at some point. I should've checked `config.go`. This will be resolved.

File: main.go (hunk 2 above)
Comment (MrSaints) [reply main.go]: Done!

File: main.go (hunk 3 above)
Comment (MrSaints) [reply main.go]: Done as well

File: .dockerignore (hunk 4 above)
Comment (MrSaints) [reply .dockerignore]: This should be addressed

**Comment by MrSaints** [comment]
@adtac Cheers for the review. I've made the appropriate changes.

File: .dockerignore (hunk 4 above)
Comment (MrSaints) [reply .dockerignore]: For some reason, this isn't being reflected on the Github editor / viewer. But I've definitely added it.

File: Dockerfile (hunk 5 above)
Comment (adtac) [reply Dockerfile]: Agreed, thanks for the explanation.

File: .dockerignore (hunk 4 above)
Comment (adtac) [reply .dockerignore]: Looks good now, weirdly.
